extensions/aws/processors/PutKinesisStream.cpp

```diff
@@ -0,0 +1,195 @@
+  entry.SetData(aws_buffer);
+  return entry;
+}
+
+std::unordered_map<std::string, PutKinesisStream::StreamBatch> PutKinesisStream::createStreamBatches(const core::ProcessContext& context, core::ProcessSession& session) const {
+  static constexpr uint64_t SINGLE_RECORD_MAX_SIZE = 1_MB;
+  std::unordered_map<std::string, StreamBatch> stream_batches;
+  uint64_t ff_count_in_batches = 0;
+  while (ff_count_in_batches < batch_size_) {
+    std::shared_ptr<core::FlowFile> flow_file = session.get();
+    if (!flow_file) { break; }
+    const auto flow_file_size = flow_file->getSize();
+    if (flow_file_size > SINGLE_RECORD_MAX_SIZE) {
+      flow_file->setAttribute(AwsKinesisErrorMessage.name, fmt::format("record too big {}, max allowed {}", flow_file_size, SINGLE_RECORD_MAX_SIZE));
+      session.transfer(flow_file, Failure);
+      logger_->log_error("Failed to publish to kinesis record {} because the size was greater than {} bytes", flow_file->getUUID().to_string(), SINGLE_RECORD_MAX_SIZE);
+      continue;
+    }
+
+    auto stream_name = context.getProperty(AmazonKinesisStreamName.name, flow_file.get());
+    if (!stream_name) {
+      logger_->log_error("Stream name is invalid due to {}", stream_name.error().message());
+      flow_file->setAttribute(AwsKinesisErrorMessage.name, fmt::format("Stream name is invalid due to {}", stream_name.error().message()));
+      session.transfer(flow_file, Failure);
+      continue;
+    }
+
+    auto entry = createEntryFromFlowFile(context, session, flow_file);
+    if (!entry) {
+      flow_file->addAttribute(AwsKinesisErrorMessage.name, entry.error().error_message);
+      if (entry.error().error_code) {
+        flow_file->addAttribute(AwsKinesisErrorCode.name, *entry.error().error_code);
+      }
+      session.transfer(flow_file, Failure);
+      continue;
+    }
+
+    auto [stream_batch, newly_created] = stream_batches.emplace(*stream_name, StreamBatch{});
+    if (newly_created) {
+      stream_batch->second.request.SetStreamName(*stream_name);
+    }
+    stream_batch->second.request.AddRecords(*entry);
+    stream_batch->second.items.push_back(BatchItem{.flow_file = std::move(flow_file), .error = std::nullopt});
+    stream_batches[*stream_name].batch_size += flow_file_size;
+    ++ff_count_in_batches;
+
+    if (stream_batches[*stream_name].batch_size > batch_data_size_soft_cap_) {
+      break;
+    }
+  }
+  return stream_batches;
+}
+
+void PutKinesisStream::processBatch(core::ProcessSession& session, StreamBatch& stream_batch, const Aws::Kinesis::KinesisClient& client) const {
+  const auto put_record_result = client.PutRecords(stream_batch.request);
+
+  if (!put_record_result.IsSuccess()) {
+    for (auto& [flow_file, error] : stream_batch.items) {
+      error = BatchItemError{.error_message = put_record_result.GetError().GetMessage(), .error_code = std::to_string(static_cast<int>(put_record_result.GetError().GetErrorType()))};
```

Comment:
These errors (also at line 136) need to be processed by lines 149-153, no? That won't happen because of the `return` statements.

Reply:
wow nice catch, I thought I covered this with test cases... better late than never https://github.com/apache/nifi-minifi-cpp/pull/1927/commits/63f65dcb5ef61e858c0756c27ad9a6939b427468